SYCL/ESIMD/api/functional/shared_element.hpp

```diff
@@ -0,0 +1,45 @@
+//===-- shared_elements.hpp - Function that provides USM with a smart pointer.
+//      -------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file provides data struct that lets interact with USM with a smart
+/// pointer that lets avoid memory leaks.
+///
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <functional>
+#include <memory>
+#include <sycl/sycl.hpp>
+
+namespace esimd_test::api::functional {
+
+template <typename T> class shared_element {
+  std::unique_ptr<T, std::function<void(T *)>> m_allocated_data;
+
+public:
+  shared_element(sycl::queue &queue) {
+    const auto &device{queue.get_device()};
+    const auto &context{queue.get_context()};
+
+    auto deleter = [=](T *ptr) { sycl::free(ptr, context); };
+
+    m_allocated_data = std::unique_ptr<T, decltype(deleter)>(
+        sycl::malloc_shared<T>(1, device, context), deleter);
+  }
```

Comment:
```suggestion
    m_allocated_data = std::unique_ptr<T, decltype(deleter)>(
        sycl::malloc_shared<T>(1, device, context), deleter);
    assert(m_allocated_data && "USM memory allocation failed");
    *m_allocated_data = initial_value;
  }
```

Reply:
I've applied this comment [here](https://github.com/intel/llvm-test-suite/pull/762/commits/92df51c1386dfd2e52709190f04f2c7baaec113d#diff-9d2d5320f066e81e77b17efb5f2888df4d6d5e653f591eb2e9d39f34c267172cR37-R38)